CODING_CONVENTIONS_C++.md

```diff
@@ -0,0 +1,369 @@
+
+- Never use `typedef`; always write `using T = X` in favor of `typedef X T`.
+
+- Keywords are always followed by a whitespace: `if (...)`, `template <...>`,
+  `while (...)`, etc.
+
+- Leave a whitespace after `!` to make negations easily recognizable:
+
+  ```cpp
+  if (! sunny())
+      stay_home();
+  else
+      go_outside();
+  ```
+
+- Opening braces belong to the same line:
+
+  ```cpp
+  void foo() {
+      // ...
+  }
+  ```
+
+- Use standard order for readability: C standard libraries, C++ standard
+  libraries, other libraries, (your) RIOT headers:
+
+  ```cpp
+  // some .hpp file
+
+  #include <sys/types.h>
+
+  #include <vector>
+
+  #include "3rd/party.h"
+
+  #include "riot/fwd.hpp"
+  ```
+
+  RIOT includes should always be in doublequotes, whereas system-wide
+  includes in angle brackets. In a `.cpp` file, the implemented header always
+  comes first and the header `riot/config.hpp` can be included second if you
+  need platform-dependent headers.
+
+- When declaring a function, the order of parameters is: outputs, then inputs.
+  This follows the parameter order from the STL.
+
+- Protect single-argument constructors with `explicit` to avoid implicit conversions.
+
+## Naming
+
+- All names except macros and template parameters should be
+  lower case and delimited by underscores.
+
+- Template parameter names should be written in CamelCase.
+
+- Types and variables should be nouns, while functions performing an action
+  should be "command" verbs. Classes used to implement metaprogramming
+  functions also should use verbs, e.g., `remove_const`.
+
+- Private and protected member variables use the suffix `_` while getter *and* 
+  setter functions use the name without suffix:
+
+  ```cpp
+  class person {
+  public:
+      const std::string& name() const {
+          return name_
+      }
+
+      void name(const std::string& new_name) {
+          name_ = new_name;
+      }
+
+  private:
+      std::string name_;
+  };
+  ```
+
+- Use `T` for generic, unconstrained template parameters and `x`
+  for generic function arguments. Suffix both with `s` for
+  template parameter packs:
+
+  ```cpp
+  template <class... Ts>
+  void print(const Ts&... xs) {
+      // ...
+  }
+  ```
+
+
+## Headers
+
+- Each `.cpp` file has an associated `.hpp` file.
+  Exceptions to this rule are unit tests and `main.cpp` files.
+
+- Each class has its own pair of header and implementation
+  files and the relative path for the header file is derived from its full name.
+  For example, the header file for `riot::example::my_class` of module
+  `my_module` is located at `path/to/my_module/riot/example/my_class.hpp`
+  and the source file at `path/to/my_module/my_class.cpp`.
+
+- All header files should use `#define` guards to prevent multiple inclusion.
+  The symbol name is `<RELATIVE>_<PATH>_<TO>_<FILE>_HPP`.
+
+- Do not `#include` when a forward declaration suffices.
+
+- Each library component must provide a `fwd.hpp` header providing forward
```

Comment:
This is directly taken from the CAF coding style but makes IMHO not really sense. The standard way to include all headers of a library `xyz` with only one include statement is to define a header file `xyz/xyz.hpp`. This rule should be revised.